README.md
# NotARook-ie
## My Attempt at a Chess Engine written in C

I started playing chess recently, so why not keep my skills in C up to date by writing a chess engine?

### Overview
This engine uses [alpha-beta search with null-move pruning implemented via negamax](https://en.wikipedia.org/wiki/Alpha%E2%80%93beta_pruning) for move selection as well as [IDDFS](https://en.wikipedia.org/wiki/Iterative_deepening_depth-first_search) to speed things up and to provide a decently good move if the algorithm is interrupted before it can calculate the best possible move. It also uses the [killer heuristic/history heuristic](https://en.wikipedia.org/wiki/Killer_heuristic) to help speed up the search for best moves.
This engine is compatible with xboard using the [xboard protocol](https://www.gnu.org/software/xboard/engine-intf.html) or other GUIs that use the <a href="http://wbec-ridderkerk.nl/html/UCIProtocol.html" target="_blank" style="text-decoration:none;">UCI protocol</a> for chess. It can also be played in command line mode without any need for an external GUI.

The engine uses the same board representation as the <a href="https://www.chessprogramming.org/Cray-1" target="_blank" style="text-decoration:none;">Cray-1 supercomputer</a> where the board is 120 squares. The top and bottom have two buffer
rows, and each side of the board has one buffer row. This is to ensure that when calculating all possible moves, null moves
are accounted for and the engine doesn't just try to move a piece off of the board. The code itself is full of comments explaining why things are the way that they are to hopefully help those needing inspiration for their own engines.

See the [issues tab](https://github.com/nate-browne/NotARook-ie/issues) to see what kinds of improvements are on the way in the future.

Also note that this engine comes bundled with a ton of opening books in the form of Polyglot books (found in the `books/` directory). See the `README.md` file instead of the `books/` directory for more information.

### Setup and Usage

#### XBoard/WinBoard setup
<details>

Here are the steps:
1. Download xboard or winboard
    * on macOS, type `brew install xboard`
        * After installing on macOS, open the program (`xboard &`) then open the sound options (Engine -> Sound) and change the sound program to `afplay`
        * Next, go to Engine -> Time Control and under the incremental tab set "Initial time" to 10 min and set "Increment or max (sec/move)" to 5 seconds to set up a 10 + 5 game
    * On linux, check out [the xboard download page](https://www.gnu.org/software/xboard/#download).
    * For windows, [check here](https://www.gnu.org/software/xboard/#tag-A1). Fair warning that this engine may or may not run well (or at all, really) on windows (haven't tested it).
2. Clone the repo
3. Type `make` to build the program.
4. To play against the engine, run `xboard -cp -fcp "bin/notarookie <path to openings book to use (optional)>" &`.
    * This will start the engine with you playing the white pieces.
5. To have the engine play itself, run `xboard -fcp "bin/notarookie <path to openings book to use (optional)>" -scp bin/notarookie <path to openings book to use (optional)> &`.
6. To play another engine, run the above command but change the part after `-scp` to be the path to the other engine's executable.
7. Once in xboard, you can use the "Mode" tab to switch if the engine is playing the white pieces or the black pieces.
</details>

#### Arena/PyChess (UCI protocol setup)
Watch [this video](https://www.youtube.com/watch?v=7WUN1dgUEmw) to see how to set up the engine to run in Arena chess.
For PyChess, you can [download here](https://pychess.github.io/download/).

#### Command line setup
<details>

1. Clone the repo
2. Type `make` to build the program.
3. Run the program with `bin/notarookie <path to openings book to use (optional)>`
4. Type `console` to go to command line mode
5. Set up the game
    * First, type `new` to initialize the board
    * Type `nopost` to hide the engine's thought process from you
    * You can now set the depth (how far ahead you want the engine to think), the time (how many seconds you want the engine to be able to think), or both to limit the engine's output time. If neither are set, the engine will calculate till depth 64, however long that may take (could be a while!)
        * depth is set with `depth x` where `x` is an integer
        * time is set with `time x` where `x` is an integer in seconds
        * you can type `view` to see the current settings
    * To play the white pieces, type a move. To play black, type `go` to tell the engine to play a move first, then play your move.
        * moves are entered in long algebraic notation:
            * "1. d4" is written as "d2d4"
            * "O-O" is written as "e1g1" (for white)
            * captures are written by just moving the piece to the square
            * promotion is written as "e7e8q" (for promotion to queen)
        * The engine will automatically respond to your move
    * You can print the current board state with `print`

**WARNING**: The engine in console mode is somewhat difficult to read since it displays everything in text on screen. It may help to have a board open (either physically or on chess.com/lichess.org) as well to make the move you see for better readability, but make sure that you _turn off_ engine evaluation on those online tools so that you're not cheating :)
</details>


### Acknowledgements
Shoutout to [Bluefever software](https://www.youtube.com/user/BlueFeverSoft) for providing a basic guide to the steps for implementing a chess engine over on YouTube. It really helped keep me from floundering around trying to guess on what task to do next.

Shoutout to [@sccontre](http://github.com/sccontre) for giving me the name for this engine.


CHANGELOG.md
This project adheres to [semantic versioning](https://semver.org/).
# 2.0.3 (2022-05-02)
***
### Release Highlights
* Fixed compile error caused by newest varion of GCC/Clang that prevented compilation in optimized mode due to an extraneous warning flag.
# 2.0.2 (2021-02-27)
***
### Release Highlights
* Fixed a bug involving pawn move generation regarding quiet moves vs capture moves.
* Engine is better at playing bullet.
    * In these time controls, the engine will only "think" for one second maximum when it finds a book move to play as opposed to the 2-4 second "think" time when playing blitz/rapid/classical/correspondence.

# 2.0.1 (2021-02-19)
***
### Release Highlights
* Cleaned up the code and added comments. Should see slight performance boost where in
10 + 5 games the engine will get to a depth of ~7-9 before it plays a move on my 2017 MacBook Pro.
Stronger computers should get to a deeper depth.

# 2.0.0 (2021-02-18)
***
### Release Highlights
* Engine now comes with and can use Polyglot opening books.
README instructions are updated and a bunch of books come bundled with the program.

# 1.3.0 (2021-02-18)
***
### Release Highlights
* Added ability for engine to print moves in standard algebraic notation
when playing in console mode.

# 1.2.2 (2021-02-14)
***
### Release Highlights
* Fixed segfault in `evaluate.c` caused from adding king evaluation.
* Updated engine's response to the `force` command.

# 1.2.1 (2021-02-14)
***
### Release Highlights
* Fixed bug in `engine.c` regarding the `quit` variable to avoid undefined behavior.

# 1.2.0 (2021-02-14)
***
### Release Highlights
* Added null move pruning to speed up searching.
  * New null move creation/takeback functions added to `makemove.c`
  * Function defs added to `functions.h`.
  * New code used in `search.c`

# 1.1.1 (2021-02-14)
***
### Release Highlights
* Fixed bug in `evaluate.c` regarding incorrect metrics
used to evaluate king safety.
* Added bonus for keeping both bishops in the game/removing
both of the opponent's bishops.

# 1.1.0 (2021-02-14)
***
### Release Highlights
* Added material draw recognition. This recognition is for
draws that are draws no matter what, not just theoretical draws.
* Updated/added piece evaluation for queens, pawns, rooks, kings.

# 1.0.1 (2021-02-13)
***
### Release Highlights
* Added check evaluation so that being in check causes us to
increase the depth, hopefully allowing us to find a move that gets us out of check.

# 1.0.0 (2021-02-12)
***
### Release Highlights
* Finally released version 1.0.0! This build is pretty unoptimized, but will still play a decent game.


notarook-ie/bboard.c
/**
 * All functions pertaining to operations on bit boards live in this file.
 * Bitboards are used for concice representation of the entire chessboard
 * within one unsigned 64 bit integer.
 * See https://en.wikipedia.org/wiki/Bitboard for more info.
 *
 * Lots of this stuff comes from the chess programming wiki, in case
 * you're interested. Specifically, from here:
 * https://www.chessprogramming.org/Looking_for_Magics
 */

#include "constants.h"

// why this order? look it up
// I just kind of said "...okay" and used it
static const int32_t BIT_TABLE[64] = {
  63, 30, 3, 32, 25, 41, 22, 33, 15, 50, 42, 13, 11, 53, 19, 34, 61, 29, 2,
  51, 21, 43, 45, 10, 18, 47, 1, 54, 9, 57, 0, 35, 62, 31, 40, 4, 49, 5, 52,
  26, 60, 6, 23, 44, 46, 27, 56, 16, 7, 39, 48, 24, 59, 14, 12, 55, 38, 28,
  58, 20, 37, 17, 36, 8
};

/**
 * Function that returns the index of the LSB in the bitboard that is set
 * to 1, while setting said index to 0
 * This is some more bit manipulation magic that I can't be bothered to
 * properly explain; hopefully the code is straightfoward enough?
 */
int32_t pop_bit(uint64_t *board_ptr) {
  uint64_t bit = *board_ptr ^ (*board_ptr - 1);

  uint32_t fold = (uint32_t)((bit & 0xFFFFFFFF) ^ (bit >> 32));

  *board_ptr &= (*board_ptr - 1);

  return BIT_TABLE[(fold * MAGIC_BIT_NUM) >> 26];
}

/**
 * Function for counting the number of set bits in a given bitboard
 * (but you probably already could tell that, from the name)
 */
int32_t count_bits(uint64_t board) {
  int32_t result;

  // weird looking, but this pretty much decrements one from the board
  // in a loop that does nothing else
  for(result = 0; board; result++, board &= board - 1);

  return result;
}
